api/utils/keys/piv/yubikey.go

```diff
@@ -41,25 +39,88 @@ import (
 	"github.com/gravitational/teleport/api/utils/retryutils"
 )
 
-const (
-	// PIVCardTypeYubiKey is the PIV card type assigned to yubiKeys.
-	PIVCardTypeYubiKey = "yubikey"
-)
+// YubiKey is a specific YubiKey PIV card.
+type YubiKey struct {
+	// conn is a shared YubiKey PIV connection.
+	//
+	// PIV connections claim an exclusive lock on the PIV module until closed.
+	// In order to improve connection sharing for this program without locking
+	// out other programs during extended program executions (like "tsh proxy ssh"),
+	// this connections is opportunistically formed and released after being
+	// unused for a few seconds.
+	*sharedPIVConnection
+	// serialNumber is the YubiKey's 8 digit serial number.
+	serialNumber uint32
+	// version is the YubiKey's version.
+	version piv.Version
+}
 
-// Cache keys to prevent reconnecting to PIV module to discover a known key.
-//
-// Additionally, this allows the program to cache the key's PIN (if applicable)
-// after the user is prompted the first time, preventing redundant prompts when
-// the key is retrieved multiple times.
-//
-// Note: in most cases the connection caches the PIN itself, and connections can be
-// reclaimed before they are fully closed (within a few seconds). However, in uncommon
-// setups, this PIN caching does not actually work as expected, so we handle it instead.
-// See https://github.com/go-piv/piv-go/issues/47
-var (
-	cachedKeys   = map[piv.Slot]*hardwarekey.PrivateKey{}
-	cachedKeysMu sync.Mutex
```

Comment:
@bl-nero This cache is dropped in this commit - ecebdca2ab9f5e0c3ebabb7fe6346715114978e1